src/trace_ng/log.zig

```diff
@@ -0,0 +1,672 @@
+        }
+
+        pub fn logfWithFields(self: *Self, level: Level, comptime fmt: []const u8, args: anytype, fields: anytype) void {
+            if (@intFromEnum(level) > @intFromEnum(self.max_level)) {
+                // noop
+                return;
+            }
+            const maybe_scope = blk: {
+                if (scope) |s| {
+                    break :blk @typeName(s);
+                } else {
+                    break :blk null;
+                }
+            };
+
+            const log_message = logfmt.createLogMessage(
+                &self.recycle_fba,
+                self.max_buffer,
+                level,
+                maybe_scope,
+                null,
+                fields,
+                fmt,
+                args,
+            );
+            const writer = self.log_msg.?.writer();
+            logfmt.writeLog(writer, log_message) catch @panic("Failed to write log");
+        }
+    };
+
+    return union(LogKind) {
+        const Self = @This();
+        standard: StanardErrLogger,
+        testing: TestingLogger,
+        noop: void,
+        pub fn init(config: Config) Self {
+            switch (config.kind) {
+                .standard => {
+                    return .{ .standard = StanardErrLogger.init(.{
+                        .allocator = config.allocator,
+                        .exit_sig = config.exit_sig,
+                        .max_level = config.max_level,
+                        .max_buffer = config.max_buffer,
+                    }) };
+                },
+                .testing, .noop => {
+                    return .{ .testing = TestingLogger.init(.{
+                        .allocator = config.allocator,
+                        .exit_sig = config.exit_sig,
+                        .max_level = config.max_level,
+                        .max_buffer = config.max_buffer,
+                    }) };
+                },
+            }
+        }
+
+        pub fn deinit(self: *Self) void {
+            switch (self.*) {
+                .standard => |logger| {
+                    var standard = logger;
+                    standard.deinit();
+                },
+                .testing => |logger| {
+                    var test_logger = logger;
+                    test_logger.deinit();
+                },
+                .noop => {},
+            }
+        }
+
+        pub fn spawn(self: *Self) void {
+            switch (self.*) {
+                .standard => |*logger| {
+                    logger.spawn();
+                },
+                .noop, .testing => {},
+            }
+        }
+
+        pub fn unscoped(self: *Self) *UnscopedLogger {
+            return @ptrCast(self);
+        }
+
+        pub fn withScope(self: *Self, comptime new_scope: anytype) *StandardLogger(new_scope) {
+            return @ptrCast(self);
+        }
+
+        pub fn log(self: *Self, level: Level, message: []const u8) void {
```

Comment:
I think in this case it is fine if we deviate from the existing usages (as we have actually done already in some part of the API)...exposing public methods for each log level would blow up the method counts (4 to 16) and it would mostly be repetitive.

Reply:
I think it would be more ergonomic and intuitive to have separate method for each log level. It's how all the popular loggers works in every language I've used.

I understand that large amounts of repetitive code can be seen as clutter, making it difficult to read and maintain code. But I think we can avoid it causing any actual problems, as long as the code is organized and the pattern is clear and simple with minimal boilerplate.

You could have a single file dedicate to defining only the logger interface. The file would just contain a single struct definition. That struct would only need 20 methods. 4 are the `log*` methods you currently have. The 16 additional methods would each have only a single line of code. The pattern would be clear and it wouldn't get in the way of anything else if that's the only thing in the file.

Reply:
> It's how all the popular loggers work in every language I've used

Then you have not used the tracing crate? :) https://docs.rs/tracing/latest/tracing/ (I think it also supports the individual methods but having the level as a param is also idiomatic as that is what is used in the examples)

Reason I was hesitant before is exactly what you also echoed "large amounts of repetitive code can be seen as clutter,"  - But I can update it, if that is indeed the requirements.

Reply:
I'm not opposed to having public methods that can be used as `log(.info`. In some cases I think those are actually more useful than `info` style methods, when the level needs to be dynamic. I'm advocating for publicly exposing *both* approaches in the api, like the tracing crate.

Let's say you're a consumer of the library and both options are available. You want to log an info message, and you have the choice between calling `log(.info` and `info`. Which one would you use? Personally I would use the info method in most cases, since it's easier to read and write. For me, that feels like a compelling reason to include it in the api.

> But I can update it, if that is indeed the requirements.

The requirements are for us to decide. I hope I didn't give the impression that it needs to strictly be whatever I say. I just wanted to share my reasoning in hope that you might agree with my conclusion. If we can't reach a consensus, we should loop in some others to get a fresh perspective.

Reply:
> I'm advocating for publicly exposing both approaches in the api, like the tracing crate.

Yup. I definitely agree with this. My initial hesitation was that there would be a blow up of the methods...but at the end of the day it turned out to be not that bad :)

> I hope I didn't give the impression that it needs to strictly be whatever I say. I just wanted to share my reasoning in hope that you might agree with my conclusion.

Nope. It did not come across like that. Although I'll say I also don't mind differing requirements spec to older contributors, given I am just start contributing, hence I don't have a lot of the context that may influence certain decisions.